### PR TITLE
Support 'let open' in class and class type expressions

### PR DESCRIPTION
**Pull request by alainfrisch**
Add support for:

````ocaml
class c =
  let open M in
  ...

class type ct =
  let open M in
  ...
````

References:

https://caml.inria.fr/mantis/view.php?id=6271
https://caml.inria.fr/mantis/view.php?id=7529

@garrigue I'm not entirely satisfied with the type-checking of open in classes, where the open is performed on both environments val_env/met_env (with a hack to share the usage marker).  Do you see a better way?

[EDIT: now ready for review]

### CODE REVIEW

File: typing/typedtree.mli
Comment (trefis) [inline typing/typedtree.mli:271]: The `Env.t` parameter never appears to be read anywhere.

File: typing/typedtree.mli
Comment (trefis) [inline typing/typedtree.mli]: The `Env.t` parameter never appears to be read anywhere.

File: typing/typeclass.ml
Comment (trefis) [inline typing/typeclass.ml:1208]: You want to use `new_val_env` here.

**Review by trefis — CHANGES_REQUESTED** [review]

File: typing/typedtree.mli
Comment (alainfrisch) [reply typing/typedtree.mli:271]: I think it's the same as for the Texp_open case.  I assumed this could be used by external tools parsing the .cmt files.

File: typing/typeclass.ml
Comment (alainfrisch) [reply typing/typeclass.ml:1208]: I don't think so.  We record here the original env used to type-check the class expression.

Comment (trefis) [reply typing/typeclass.ml:1208]: the implementation of `Pcl_let` a few lines above seems to disagree.

Comment (trefis) [reply typing/typeclass.ml:1208]: But it is also possible that it is the handling of `Pcl_let` which is "wrong". At least I agree it is inconsistent with the way we handle `_env` fields in other places (eg. expressions)

Comment (alainfrisch) [reply typing/typeclass.ml:1208]: Indeed, but I think this is a bug.  (Pcl_fun, for instance, stores the initial val_env.)  This dates back from commit f2095623.  @garrigue : can you comment?

File: typing/typedtree.mli
Comment (trefis) [reply typing/typedtree.mli:271]: Fair point, thanks for explaining :)

File: typing/typeclass.ml
Comment (garrigue) [reply typing/typeclass.ml:1208]: My understanding was that the environment here is just there so that we can properly interpret identifiers.
So any of the two environments is ok (the new one has just more information).
At least when it was introduced, the goal was not to be able to use this information to compile source code.
If we want to do that, we should define a more precise semantics.

**Comment by garrigue** [comment]
The need to update both environments seems natural to me. It would be hard to give a semantics of `open` where you update only one of them...
As for the hack in `Env.open_signature` it has the advantage of making things explicit.
I actually find the implicit approach to usage warnings, using lots of side-effects, to be mind-boggling, and would prefer more of this.

File: typing/typeclass.ml
Comment (alainfrisch) [reply typing/typeclass.ml:1208]: > My understanding was that the environment here is just there so that we can properly interpret identifiers.

Indeed, a quick look at translclass suggest that cl_env is mostly used on Tcl_ident node.  There is also `build_class_lets`, but it deals with Tcl_let directly (and return the inner env explicitly).  The remaining use is through `Translobj.oo_wrap/Translobj.oo_add_class`, which ends up defining the `top_env` in `transl_class` (and then passed to Env.diff).  And here I'm a bit lost.

> So any of the two environments is ok (the new one has just more information).
> If we want to do that, we should define a more precise semantics.

I think the least confusing semantics would be to store the val_env "around" the class expression.  This is similar to the semantics of exp_env, pat_env, mod_env.  Do you agree?

> At least when it was introduced, the goal was not to be able to use this information to compile source code.

I'm not sure what you are referring to, but in my understanding, this more is more about supporting external tools (documentation generators, IDE support, etc) so that they can do some lookups.  It's not really about allowing to compile source code fragment (although this might be useful as well).

Comment (alainfrisch) [reply typing/typeclass.ml:1208]: Anyway, do we agree that for Pcl_open, it's ok to keep the initial val_env (as for Pcl_fun)?  The change to Pcl_let should be dealt with separately.

Comment (garrigue) [reply typing/typeclass.ml:1208]: I agree that the code for `Pcl_open` is correct.

**Review by trefis — APPROVED** [review]
Considering the previous discussion this looks OK to me.

(N.B. I don't have any opinion on the manual changes, as I don't know how the manual is organized)

**Comment by Octachron** [comment]
The manual changes look goods to me. I was thinking that using a local open inside of the examples in the class chapter of the tutorial may be nice but this feels very optional since the syntax is easily discoverable starting from the syntax for local open in expression.

**Comment by alainfrisch** [comment]
@garrigue If you agree with the PR, can you create a formal review?  As per the new rule, we need a review from someone with commit rights to merge something that touches the compiler.

**Review by garrigue — APPROVED** [review]
The change is a simple one, and easily tested. As any syntactic change, there are many files involved, but there shouldn't be any negative impact on existing code anyway.

**Pull request merged**
